src/python/pants/backend/k8s/goals/deploy.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+
+from pants.backend.docker.goals.package_image import DockerPackageFieldSet
+from pants.backend.k8s.k8s_subsystem import K8sSubsystem
+from pants.backend.k8s.kubectl_subsystem import Kubectl
+from pants.backend.k8s.target_types import (
+    K8sBundleContextField,
+    K8sBundleDependenciesField,
+    K8sBundleSourcesField,
+    K8sSourceField,
+)
+from pants.core.goals.deploy import DeployFieldSet, DeployProcess
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.engine.addresses import UnparsedAddressInputs
+from pants.engine.env_vars import EnvironmentVars, EnvironmentVarsRequest
+from pants.engine.fs import MergeDigests, Snapshot
+from pants.engine.internals.native_engine import Digest
+from pants.engine.platform import Platform
+from pants.engine.process import InteractiveProcess, Process, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    DependenciesRequest,
+    HydratedSources,
+    HydrateSourcesRequest,
+    SourcesField,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DeployK8sBundleFieldSet(DeployFieldSet):
+    required_fields = (
+        K8sBundleSourcesField,
+        K8sBundleContextField,
+        K8sBundleDependenciesField,
+    )
+    sources: K8sBundleSourcesField
+    context: K8sBundleContextField
+    dependencies: K8sBundleDependenciesField
+
+
+@dataclass(frozen=True)
+class KubectlApply:
+    paths: tuple[str, ...]
+    input_digest: Digest
+    platform: Platform
+    env: FrozenDict[str, str] | None = None
+    context: str | None = None
+
+
+@rule(desc="Run k8s deploy process", level=LogLevel.DEBUG)
+async def run_k8s_deploy(
+    field_set: DeployK8sBundleFieldSet,
+    kubectl: Kubectl,
+    k8s_subsystem: K8sSubsystem,
+    platform: Platform,
+) -> DeployProcess:
+    context = field_set.context.value
+    if context is None:
+        raise ValueError(f"Missing `{K8sBundleContextField.alias}` field")
```

Comment:
we could make this more helpful by including the FS's address with `field_set.address.spec`

Reply:
Good suggestion, thanks https://github.com/pantsbuild/pants/pull/21796/commits/ed0be1f91eaf2436ea8a9ae587faafd1b3675b2e